feat(snapshot,restore): support for local snapshot

mynktl opened the pull request:
Ref: https://github.com/openebs/openebs/issues/2844

Changes:
- Added support for local snapshot restore of CStor volume.

Sample volumeSnapshotLocation:
```
spec:
  config:
    local: "true"
  provider: openebs.io/cstor-blockstore
```

vishnuitta commented on pkg/cstor/api_service.go:
```diff
@@ -116,3 +119,103 @@ fetchip:
+
+	bkpSpec := &v1alpha1.CStorBackupSpec{
+		BackupName: bname,
+		VolumeName: vol.volname,
```
is vol.volname already populated?

mynktl replied:
yes. By `GetVolumeID`

vishnuitta commented on README.md:
````diff
@@ -62,6 +62,17 @@ spec:
     region: <AWS_REGION>
 ```
+
+For local snapshot of CStor volume through velero, Refere `example/06-local-volumesnapshotlocation.yaml` to configure `VolumeSnapshotLocation`.
````
to be consistent with 'cloud backup', should we call this as 'local backup'? before changing, lets get opinion from @kmova as well

mynktl replied:
That may create confusion with local copy of the snapshot.

kmova replied:
For creating local backup using cStor Volume snapshot feature, refer `example/06-local-volumesnapshotlocation.yaml` to configure `VolumeSnapshotLocation`.

vishnuitta commented on README.md:
```diff
@@ -102,6 +113,13 @@ Once the backup is completed you should see the backup marked as `Completed`.
 
 
 ### Creating a restore from backup
+#### Creating a restore from local backup/snapshot
```
lets put limitations as well for restore from local snapshot if any like 'storageclass can't have WaitForFirstConsumer'

vishnuitta replied:
and `restore from local backup need to be done from same cluster only where local backups are done, unlike restore of remote backups which can be done to any cluster.`

mynktl replied:
ok

vishnuitta commented on pkg/cstor/cstor.go:
```diff
@@ -506,3 +499,35 @@ func (p *Plugin) SetVolumeID(unstructuredPV runtime.Unstructured, volumeID strin
+
+	// To support namespace re-mapping for cloud-snapshot remove below check
+	if p.local {
+		// Let's rename PV if already created
```
few comments on what cases it can exists and what cases it doesn't would be good to have.
With new 'p.local' conditions, I guess - more such comments are good to have in code.

mynktl replied:
ok

vishnuitta commented on pkg/cstor/pv_operation.go:
```diff
@@ -0,0 +1,117 @@
+		return "", errors.Wrapf(err, "Error checking if PV with same name exist")
+	}
+
+	if pv.Spec.ClaimRef == nil {
```
are we ok to take this PV if not claimed?
are we supposed to check on some annotations as well like the ones that are added when PV is created for restore from local backup?

vishnuitta replied:
or from its name if it contains 'cstor-clone-'

mynktl replied:
If PV is already created, but not claimed, then relevant CStor resources also should exist. So we can't use. I'll remove this check. Regarding Annotation, we don't need to check, Since we will be restoring from backup only.

vishnuitta commented on README.md:
```diff
@@ -102,6 +113,13 @@ Once the backup is completed you should see the backup marked as `Completed`.
+#### Creating a restore from local backup/snapshot
+To restore local snapshot, run the following command:
+
+`velero restore create --from-backup backup_name --restore-volumes=true --namespace-mappings source_ns:destination_ns`
```
does namespace-mappings should work for restore of remote backup as well right?
or is this for next PR?

mynktl replied:
no. It is not supported. I've added limitation. It requires handing of incremental backups.

vishnuitta commented on pkg/cstor/cstor.go:
```diff
@@ -427,57 +431,32 @@ func (p *Plugin) CreateVolumeFromSnapshot(snapshotID, volumeType, volumeAZ strin
 	if err != nil {
-		p.Log.Errorf("Error during JSON marshal : %s", err.Error())
-		return "", err
+		return "", errors.Wrapf(err, "Failed to read PVC")
```
better to print names also.. not only here, in createPVC also..
but, in order NOT to dilute this PR, it can taken for next PR..

mynktl replied:
ok.

vishnuitta commented on pkg/cstor/cstor.go:
```diff
@@ -506,3 +499,35 @@ func (p *Plugin) SetVolumeID(unstructuredPV runtime.Unstructured, volumeID strin
 }
+
+func (p *Plugin) getVolInfo(volumeID, snapName string) (*Volume, error) {
+	// To support namespace re-mapping for cloud snapshot,
```
ok.. got to see this comment.. so, still some work is pending for namespace re-mapping for cloud snapshots

mynktl replied:
namespace re-mapping for cloud snapshot is out of scope for this PR. We will add support for this in near future.

mynktl commented:
Crated an issue to track test case https://github.com/openebs/velero-plugin/issues/54.

kmova commented on README.md:
```diff
@@ -50,7 +50,7 @@ Run the following command to install development image of OpenEBS velero-plugin
 
 ## Configuring snapshot location
-To take a backup of CStor volume through Velero, configure `VolumeSnapshotLocation` with provider `openebs.io/cstor-blockstore`. Sample YAML file for volumesnapshotlocation can be found at `example/06-volumesnapshotlocation.yaml`.
+To take a cloud backup of CStor volume through Velero, configure `VolumeSnapshotLocation` with provider `openebs.io/cstor-blockstore`. Sample YAML file for volumesnapshotlocation can be found at `example/06-volumesnapshotlocation.yaml`.
```
To take remote backup of cStor volume snapshot to cloud or S3 compatible storage, configure `VolumeSnapshotLocation` with provider `openebs.io/cstor-blockstore`. 

Can you provide samples for popular use cases like: 
- Minio
- AWS S3
- GCP buckets. 

They could be different commented out YAMLs in the same snapshotlocation example yaml.

kmova replied:
Throughout this document - mention remote/cloud backup and local backup.

mynktl replied:
ok

kmova commented on example/06-local-volumesnapshotlocation.yaml:
```diff
@@ -0,0 +1,25 @@
+apiVersion: velero.io/v1
+kind: VolumeSnapshotLocation
+metadata:
+  name: <LOCATION_NAME>
```
Add a comment above on example valid location name string.

mynktl replied:
ok

kmova commented on example/06-local-volumesnapshotlocation.yaml:
```diff
@@ -0,0 +1,25 @@
+spec:
+  provider: openebs.io/cstor-blockstore
+  config:
+    namespace: <OPENEBS_NAMESPACE>
```
it is ok to repeat the comment here on what the namespace means.

mynktl replied:
ok

kmova commented on example/06-volumesnapshotlocation.yaml:
```diff
@@ -24,6 +24,6 @@ spec:
     prefix: <PREFIX_FOR_BACKUP_NAME>
     backupPathPrefix: <PREFIX_FOR_BACKUP_PATH>
-    provider: <GCP_OR_AWS>
+    provider: <gcp_OR_aws>
```
If there are any know gotchas around permissions or how the buckets are created, please add them as comments in this YAML.

mynktl replied:
ok

kmova commented:
Gave some minor comments @mynktl - also can you confirm if a test for local backup and restore can be added in this PR or planned for another PR?

mynktl commented:
Hi @kmova 
I have addressed the review comments. For test case We can add it in another PR. As of now I've created an issue to track this. https://github.com/openebs/velero-plugin/issues/54.
